### PR TITLE
Calculate brut force QoI for wave example

### PR DESCRIPTION
**Pull request by am-kaiser**
With the brut force algorithm the return value (see documentation) is calculated for num_estimates runs. There are two options on how to calculate the return value either quantiles are chuncks of return year size can be used.

Notes:
The variance is returned. Therefore, we do not need any further uncertainty quantification.

### CODE REVIEW

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 1]
```diff
@@ -1 +1,170 @@
...
+        with results_path.open() as fp:
+            results = json.load(fp)
+            samples = torch.tensor(results["samples"])
+
+    # to reduce run time for testing
+    year_return_value = 10
+
+    # make any additional samples required
+    if len(samples) < num_estimates:
+        if brut_force_type == "quantile":
```
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/brute_force.py:77]: perhaps we wan to just use the chunk method - based on the discussion with odin we could basically remove quantile brute force from our code base. I'm happy to keep `quantile_brute_force_calc` around for now - but maybe its simpler in the rest of the code if its removed given at this point we don't think we should be using it.

File: examples/crest_heights_north_sea/problem/brute_force.py
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/brute_force.py:1]: lets review how we use the dataloader here. Might also simplify thing so we can use `period_len` to represent `n_sea-states-in_year` and `n_sea_states_in_period`

File: examples/crest_heights_north_sea/problem/data/data_generator.py [hunk 2]
```diff
@@ -1,26 +1,40 @@
...
+
+    Returns:
+        significant wave height and peak wave period
+
+    """
+    weib_prms = (1.2550, 0.5026, 1.9536)
+    lognorm_prms = (1.557, 0.403, 0.408, 0.005, 0.137, 0.454)
+
+    n_ss = n_sea_states_in_year * n_years_in_period
+    return sample_seastates(n_ss, weib_prms, lognorm_prms, hslim=7.5)
```
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/data/data_generator.py:28]: Lets use the seed functionality so this reproducible.

File: examples/crest_heights_north_sea/problem/problem.py [hunk 3]
```diff
@@ -79,10 +79,62 @@ def make_exp() -> Experiment:
...
 # dataset and dataloader
-dataset: Dataset[NDArray[np.float64]] = MinimalDataset(np.load("data/long_term_distribution.npy"))
-
+dataset: Dataset[NDArray[np.float64]] = MinimalDataset(
+    np.load(f"data/long_term_distribution_{_n_years_in_period}_years_no_hslim.npy")
+)
 
 dataloader = DataLoader(dataset, batch_size=256, shuffle=True)
 
+# %%
```
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/problem.py:88]: all this should probably be at the bottom of `brute_force.py`

**Review by swinter1 — CHANGES_REQUESTED** [review]
This doesn't have to be the same as what we have got in Demo2d, but lets have a chat about what parts would be useful to ring across. lets find a time to sit together and talk through the dataloader use.

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 4]
```diff
@@ -1 +1,170 @@
...
+    """Generate the file name from the period length."""
+    return f"n_sample_per_period_{period_length}"
+
+
+def collect_or_calculate_results(
+    dataloader: DataLoader[tuple[Tensor, ...]],
+    n_sea_states_in_year: int,
+    n_sea_states_in_period: int,
+    num_estimates: int = 2_000,
+    brut_force_type: str = "quantile",
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done (removed quantile method)

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 5]
```diff
@@ -1 +1,170 @@
...
+def chunck_brute_force_calc(
+    dataloader: DataLoader[tuple[Tensor, ...]],
+    n_sea_states_in_period: int,
+    n_sea_states_in_year: int = 2922,
+    num_estimates: int = 2_000,
+    year_return_value: int = 100,
+) -> Tensor:
+    """Calculate the QOI by brute force by splitting the data into year_return_value chunck.
+
+    Note: At the meeting with Odin on the 22.04.25 it was decided that this is the appropriate method.
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done (removed quantile method)

File: examples/crest_heights_north_sea/problem/brute_force.py (hunk 1 above)
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done (removed quantile method)

File: examples/crest_heights_north_sea/problem/brute_force.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 6]
```diff
@@ -1 +1,170 @@
...
+
+    samples = torch.tensor([])
+
+    if results_path.exists():
+        with results_path.open() as fp:
+            results = json.load(fp)
+            samples = torch.tensor(results["samples"])
+
+    # to reduce run time for testing
+    year_return_value = 10
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 7]
```diff
@@ -1 +1,170 @@
...
+        n_sea_states_in_year: number of sea states in one year
+        num_estimates: The number of brute force estimates of the QoI. A new period is drawn for each estimate.
+        year_return_value: year for the return value
+
+    Returns:
+        The QoI values calculated for each period. Shape (num_estimates,)
+    """
+    return_value = zeros(num_estimates)
+    for i in tqdm.tqdm(range(num_estimates)):
+        simulator_samples: np.ndarray[tuple[int,], Any] = max_crest_height_simulator_function(dataloader.dataset.data)  # type: ignore  # noqa: PGH003
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done

File: examples/crest_heights_north_sea/problem/brute_force.py
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py:1]: done

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 8]
```diff
@@ -1 +1,170 @@
...
+        n_sea_states_in_period: number of sea states in whole period
+        num_estimates: The number of brute force estimates of the QoI. A new period is drawn for each estimate.
+        brut_force_type: Choose how the QoI shall be estimated.
+
+    Returns:
+        The QoI values calculated for each period. Shape (num_estimates,)
+    """
+    # Calculate the number of samples that create a single period of the ERD
+    period_length = dataloader.dataset.data.shape[0]  # type: ignore[attr-defined]
+    results_path = _results_dir / f"{_result_file_name(period_length)}_{brut_force_type}.json"
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 9]
```diff
@@ -1 +1,170 @@
...
+        return ResultsObject(statistics=statistics, samples=samples.tolist())
+
+
+def _result_file_name(period_length: int) -> str:
+    """Generate the file name from the period length."""
+    return f"n_sample_per_period_{period_length}"
+
+
+def collect_or_calculate_results(
+    dataloader: DataLoader[tuple[Tensor, ...]],
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 10]
```diff
@@ -1 +1,170 @@
...
+    year_return_value: int = 100,
+) -> Tensor:
+    """Calculate the QOI by brute force by splitting the data into year_return_value chunck.
+
+    Note: At the meeting with Odin on the 22.04.25 it was decided that this is the appropriate method.
+
+    Args:
+        dataloader: The dataloader to use to get the environment samples.
+             - Each batch should have shape (batch_size, d)
+             - The sum of the batch sizes returned by iterating through the dataloader should be a period length
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py:119]: done

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 11]
```diff
@@ -1 +1,170 @@
...
+        The QoI values calculated for each period. Shape (num_estimates,)
+    """
+    maxs = []
+    for _ in tqdm.tqdm(range(num_estimates)):
+        chunck_indices = np.append(
+            np.arange(0, n_sea_states_in_period, year_return_value * n_sea_states_in_year), n_sea_states_in_period
+        )
+
+        for idx, chunck_index in enumerate(chunck_indices[1:]):
+            samples = dataloader.dataset.data[chunck_indices[idx] : chunck_index, :]  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done

File: examples/crest_heights_north_sea/problem/problem.py (hunk 3 above)
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/problem.py]: done

File: examples/crest_heights_north_sea/problem/data/data_generator.py (hunk 2 above)
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/data/data_generator.py]: done

File: examples/crest_heights_north_sea/problem/problem.py [hunk 12]
```diff
@@ -79,10 +80,22 @@ def make_exp() -> Experiment:
 
 # %%
 # dataset and dataloader
-dataset: Dataset[NDArray[np.float64]] = MinimalDataset(np.load("data/long_term_distribution.npy"))
-
+dataset: Dataset[NDArray[np.float64]] = MinimalDataset(
+    np.load(f"data/long_term_distribution_{_n_years_in_period}_years.npy")
+)
 
 dataloader = DataLoader(dataset, batch_size=256, shuffle=True)
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/problem.py:87]: done

File: examples/crest_heights_north_sea/problem/brute_force.py [hunk 13]
```diff
@@ -1 +1,174 @@
...
+
+    @classmethod
+    def from_samples(cls, samples: torch.Tensor) -> "ResultsObject":
+        """Create the object directly from samples."""
+        statistics = {"median": float(samples.median()), "mean": float(samples.mean())}
+        return ResultsObject(statistics=statistics, samples=samples.tolist())
+
+
+def collect_or_calculate_results(
+    n_years_in_period: int,
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/brute_force.py]: done

File: examples/crest_heights_north_sea/problem/problem.py [hunk 14]
```diff
@@ -79,10 +80,22 @@ def make_exp() -> Experiment:
 
 # %%
 # dataset and dataloader
-dataset: Dataset[NDArray[np.float64]] = MinimalDataset(np.load("data/long_term_distribution.npy"))
-
+dataset: Dataset[NDArray[np.float64]] = MinimalDataset(
```
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/problem.py]: done

File: examples/crest_heights_north_sea/problem/problem.py [hunk 15]
```diff
@@ -47,29 +44,30 @@
...
 
-N_ENV_SAMPLES_PER_PERIOD = 1000  # Arbitrary number of env samples per period
+# In ax a period refers to a time length that a single extreme response relates to
+# which is in this use case the number of sea states in the desired return period
+period_length = year_return_value * n_sea_states_in_year
 
 # %%
-# TODO(@henrikstoklandberg): Find/define the bruteforce QOI for this problem and period
+# Set axtreme specific parameters
+N_ENV_SAMPLES_PER_PERIOD = 1000  # The number of simulations to run for each point in the experiment.
```
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/problem.py:65]: `N_ENV_SAMPLES_PER_PERIOD` is the legacy name we had `period_len`. Lets remove this so people don't get confused.

File: examples/crest_heights_north_sea/problem/problem.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (swinter1) [reply examples/crest_heights_north_sea/problem/problem.py]: Sorry I think I was confusing. The dataset can be kept, the dataloader should be removed.

File: examples/crest_heights_north_sea/problem/problem.py [hunk 16]
```diff
@@ -47,29 +44,30 @@
...
+num_estimates = 20  # The number of brute force estimates of the QoI. A new period is drawn for each estimate.
 
 
 # %%
-### Automatically set up your experiment using the sim, search_space, and dist defined above.
+# Automatically set up your experiment using the sim, search_space, and dist defined above.
 def make_exp() -> Experiment:
```
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/problem.py:71]: Please update this
`return make_experiment(sim, SEARCH_SPACE, DIST, n_simulations_per_point=N_ENV_SAMPLES_PER_PERIOD)`
 to this
`return make_experiment(sim, SEARCH_SPACE, DIST, n_simulations_per_point=10_000)`

I spotted this before but didn't have the chance to fix it, if you're doing the N_ENV_SAMPLES_PER_PERIOD thing perhaps you can fix this while you're at it

File: examples/crest_heights_north_sea/problem/data/long_term_distribution_1000_years.npy
Comment (swinter1) [inline examples/crest_heights_north_sea/problem/data/long_term_distribution_1000_years.npy:1]: Please remove any of these `long_term_distribtuion_....npy` files that are no longer relevant

File: examples/crest_heights_north_sea/problem/problem.py (hunk 16 above)
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/problem.py:71]: done

File: examples/crest_heights_north_sea/problem/data/long_term_distribution_1000_years.npy
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/data/long_term_distribution_1000_years.npy:1]: done

File: examples/crest_heights_north_sea/problem/problem.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (am-kaiser) [reply examples/crest_heights_north_sea/problem/problem.py]: done

**Review by swinter1 — APPROVED** [review]
